### PR TITLE
Update to latest cadvisor and containerd (v1.3.3)

### PR DESCRIPTION
**Pull request by dims**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
We use containerd 1.3.x in [cluster-api](https://github.com/kubernetes-sigs/image-builder/blob/master/images/capi/packer/config/containerd.json#L2) / [kind](https://github.com/kubernetes-sigs/kind/blob/master/images/base/Dockerfile#L26) already, it's time to move up to newer dependencies.

Need reviews for/from:

sig-windows:
- [x] github.com/Microsoft/go-winio
- [x] github.com/Microsoft/hcsshim
- [x] github.com/konsorten/go-windows-terminal-sequences

sig-node:
- [x] github.com/containerd/cgroups
- [x] github.com/containerd/console
- [x] github.com/containerd/containerd
- [x] github.com/containerd/ttrpc
- [x] github.com/godbus/dbus
- [x] github.com/google/cadvisor
- [x] github.com/mistifyio/go-zfs
- [ ] github.com/opencontainers/runc
- [x] github.com/opencontainers/runtime-spec
- [ ] github.com/opencontainers/selinux

sig-apimachinery:
- [x] github.com/gogo/protobuf
- [x] github.com/golang/groupcache
- [x] github.com/google/go-cmp
- [x] github.com/hashicorp/golang-lru
- [x] google.golang.org/genproto
- [x] ~~google.golang.org/grpc~~

sig-instrumentation:
- [x] github.com/prometheus/procfs (sig-instrumentation)
- [x] go.opencensus.io (sig-instrumentation)

Miscellaneous:
- [x] github.com/onsi/gomega (sig-testing)
- [x] github.com/vishvananda/netns (sig-network)
- [ ] golang.org/x/xerrors
- [x] google.golang.org/appengine (sig-auth)
- [ ] ~~github.com/cilium/ebpf~~
- [x] k8s.io/utils/mount (sig-storage)

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -14,10 +14,9 @@ require (
 	github.com/Azure/go-autorest/autorest/to v0.2.0
 	github.com/GoogleCloudPlatform/k8s-cloud-provider v0.0.0-20190822182118-27a4ced34534
 	github.com/JeffAshton/win_pdh v0.0.0-20161109143554-76bb4ee9f0ab
-	github.com/Microsoft/go-winio v0.4.14
-	github.com/Microsoft/hcsshim v0.0.0-20190417211021-672e52e9209d
+	github.com/Microsoft/go-winio v0.4.15-0.20190919025122-fc70bd9a86b5
```
Comment (liggitt) [inline go.mod]: is there a tagged level we can use?

File: go.mod [hunk 2]
```diff
@@ -59,76 +53,67 @@ require (
...
 	github.com/go-ozzo/ozzo-validation v3.5.0+incompatible // indirect
-	github.com/godbus/dbus v0.0.0-20181101234600-2ff6f7ffd60f // indirect
-	github.com/gogo/protobuf v1.2.2-0.20190723190241-65acae22fc9d
-	github.com/golang/groupcache v0.0.0-20160516000752-02826c3e7903
+	github.com/gogo/protobuf v1.3.1
+	github.com/golang/groupcache v0.0.0-20191227052852-215e87163ea7
 	github.com/golang/mock v1.2.0
-	github.com/google/cadvisor v0.35.0
-	github.com/google/go-cmp v0.3.0
+	github.com/google/cadvisor v0.35.1-0.20200109170520-9991c03427eb
```
Comment (liggitt) [inline go.mod]: is there a tagged level we can use?

File: go.mod (hunk 1 above)
Comment (dims) [reply go.mod]: asked them the question in https://github.com/microsoft/go-winio/issues/167

File: go.mod (hunk 2 above)
Comment (dims) [reply go.mod]: cadvisor, we always cut a fresh release before our code freeze .. ( cc @dashpole ). Also may be better to wait until the go-winio gets sorted out first before we cut another cadvisor release

Comment (dashpole) [reply go.mod]: This is fine to do for now.  I'll update it to a tagged release before the next kubernetes release.

**Comment by BenTheElder** [comment]
> We use containerd 1.3.x in cluster-api / kind already, it's time to move up to newer dependencies.

we do but ... via CRI, I wouldn't expect any k/k repo changes to be related :O

**Comment by dims** [comment]
@BenTheElder right, i am just using that as an excuse to move the goal posts here :)

**Comment by dims** [comment]
@PatrickLang - can you please approve the dependency updates related to sig-windows? (cc @kubernetes/sig-windows-misc) 
@derekwaynecarr @mattjmcnaughton @Random-Liu - can you please ok the changes for sig-node? (cc @kubernetes/sig-node-pr-reviews)
@lavalamp @liggitt - can you please look at the ones from sig-apimachinery? (cc @kubernetes/sig-api-machinery-misc)
@spiffxp @stevekuznetsov - can you please look at the one for sig-testing? (cc @kubernetes/sig-testing)  
@thockin @dcbw - can you please look at the one for sig-network? (cc @kubernetes/sig-network-pr-reviews) 
@enj @mikedanese -  can you please look at the one for sig-auth? (cc @kubernetes/sig-auth-pr-reviews)

**Comment by dims** [comment]
@brancz @piosz can you please look at the dependencies for sig-instrumentation? (cc @kubernetes/sig-instrumentation-pr-reviews)

**Comment by dims** [comment]
/assign @thockin @smarterclayton 

(for root approval)

**Comment by dims** [comment]
/hold

(for reviews)

**Comment by lavalamp** [comment]
What do you think about updating gogoproto first in a separate change for easier testing & rollback?

**Comment by lavalamp** [comment]
lgtm for github.com/hashicorp/golang-lru

**Comment by mikedanese** [comment]
go-cmp and appengine lgtm

**Comment by mattjmcnaughton** [comment]
/uncc

going to let @derekwaynecarr or @dchen1107 review this for sig-node... i'm not sure i have the full context on all aspects of the `kubelet` to know what upgrading containerd could impact.

**Comment by liggitt** [comment]
github.com/golang/groupcache lgtm

**Review by stevekuznetsov — APPROVED** [review]
`vendor/github.com/onsi/gomega/` LGTM

File: go.mod [hunk 3]
```diff
@@ -185,14 +171,19 @@ replace (
 	github.com/GoogleCloudPlatform/k8s-cloud-provider => github.com/GoogleCloudPlatform/k8s-cloud-provider v0.0.0-20190822182118-27a4ced34534
 	github.com/JeffAshton/win_pdh => github.com/JeffAshton/win_pdh v0.0.0-20161109143554-76bb4ee9f0ab
 	github.com/MakeNowJust/heredoc => github.com/MakeNowJust/heredoc v0.0.0-20170808103936-bb23615498cd
-	github.com/Microsoft/go-winio => github.com/Microsoft/go-winio v0.4.14
-	github.com/Microsoft/hcsshim => github.com/Microsoft/hcsshim v0.0.0-20190417211021-672e52e9209d
+	github.com/Microsoft/go-winio => github.com/Microsoft/go-winio v0.4.15-0.20190919025122-fc70bd9a86b5
+	github.com/Microsoft/hcsshim => github.com/Microsoft/hcsshim v0.8.7
```
Comment (PatrickLang) [inline go.mod]: Based on this - we should still be using a tagged commit ID, not version. 0.8.7 wasn't tested with Kubernetes before it was tagged. If we want to use that as a testing point I'm ok with it, but I think we should tag the commit ID so the next person updating this doesn't assume there will be a relevant tagged version. 

https://github.com/microsoft/hcsshim/issues/642#issuecomment-571772207

Comment (dims) [reply go.mod]: @PatrickLang yes, let's please use it as a testing point.

**Comment by derekwaynecarr** [comment]
sig-node updates are fine.

/approve

File: go.mod [hunk 4]
```diff
@@ -216,27 +207,36 @@ replace (
 	github.com/bradfitz/go-smtpd => github.com/bradfitz/go-smtpd v0.0.0-20170404230938-deb6d6237625
 	github.com/caddyserver/caddy => github.com/caddyserver/caddy v1.0.3
 	github.com/cenkalti/backoff => github.com/cenkalti/backoff v2.1.1+incompatible
+	github.com/census-instrumentation/opencensus-proto => github.com/census-instrumentation/opencensus-proto v0.2.1
 	github.com/cespare/prettybench => github.com/cespare/prettybench v0.0.0-20150116022406-03b8cfe5406c
 	github.com/chai2010/gettext-go => github.com/chai2010/gettext-go v0.0.0-20160711120539-c6fed771bfd5
-	github.com/checkpoint-restore/go-criu => github.com/checkpoint-restore/go-criu v0.0.0-20181120144056-17b0214f6c48 // 17b0214f6c48 is the SHA for git tag 3.11
```
Comment (dims) [inline go.mod]: This change did not stick. 2 steps forward and one step back!

**Comment by dims** [comment]
/assign @liggitt 

Jordan,
this is ready. all green and got acks from all the sigs.

File: go.mod [hunk 5]
```diff
@@ -260,11 +265,13 @@ replace (
 	github.com/flynn/go-shlex => github.com/flynn/go-shlex v0.0.0-20150515145356-3f9db97f8568
 	github.com/fogleman/gg => github.com/fogleman/gg v1.2.1-0.20190220221249-0403632d5b90
 	github.com/fsnotify/fsnotify => github.com/fsnotify/fsnotify v1.4.7
+	github.com/garyburd/redigo => github.com/garyburd/redigo v0.0.0-20150301180006-535138d7bcd7
```
Comment (liggitt) [inline go.mod]: this is a sunset repository used only by cadvisor redis storage. we should find a way to isolate storage and container integrations in cadvisor so we don't pull this into our dependency tree if we don't need to

File: go.mod [hunk 6]
```diff
@@ -334,21 +344,25 @@ replace (
...
 	github.com/hashicorp/go-syslog => github.com/hashicorp/go-syslog v1.0.0
-	github.com/hashicorp/golang-lru => github.com/hashicorp/golang-lru v0.5.1
+	github.com/hashicorp/golang-lru => github.com/hashicorp/golang-lru v0.5.3
 	github.com/hashicorp/hcl => github.com/hashicorp/hcl v1.0.0
 	github.com/heketi/heketi => github.com/heketi/heketi v9.0.1-0.20190917153846-c2e2a4ab7ab9+incompatible
 	github.com/heketi/tests => github.com/heketi/tests v0.0.0-20151005000721-f3775cbcefd6
 	github.com/hpcloud/tail => github.com/hpcloud/tail v1.0.0
 	github.com/imdario/mergo => github.com/imdario/mergo v0.3.5
 	github.com/inconshreveable/mousetrap => github.com/inconshreveable/mousetrap v1.0.0
+	github.com/influxdb/influxdb => github.com/influxdb/influxdb v0.9.6-0.20151125225445-9eab56311373
```
Comment (liggitt) [inline go.mod]: this pulls in a lot of additional transitive dependencies and is only used by cadvisor influxdb storage (which we don't use). we should find a way to isolate storage and container integrations in cadvisor so we don't pull this into our dependency tree if we don't need to

File: go.mod [hunk 7]
```diff
@@ -411,12 +427,14 @@ replace (
...
-	github.com/onsi/gomega => github.com/onsi/gomega v1.7.0
+	github.com/onsi/gomega => github.com/onsi/gomega v1.7.1
 	github.com/opencontainers/go-digest => github.com/opencontainers/go-digest v1.0.0-rc1
 	github.com/opencontainers/image-spec => github.com/opencontainers/image-spec v1.0.1
 	github.com/opencontainers/runc => github.com/opencontainers/runc v1.0.0-rc9
-	github.com/opencontainers/runtime-spec => github.com/opencontainers/runtime-spec v1.0.0
+	github.com/opencontainers/runtime-spec => github.com/opencontainers/runtime-spec v1.0.1
+	github.com/opencontainers/runtime-tools => github.com/opencontainers/runtime-tools v0.0.0-20181011054405-1d69bd0f9c39
 	github.com/opencontainers/selinux => github.com/opencontainers/selinux v1.3.1-0.20190929122143-5215b1806f52
+	github.com/pborman/uuid => github.com/pborman/uuid v0.0.0-20150824212802-cccd189d45f7
```
Comment (liggitt) [inline go.mod]: undoes https://github.com/kubernetes/kubernetes/pull/84770

File: go.mod [hunk 8]
```diff
@@ -520,6 +543,7 @@ replace (
 	gopkg.in/inf.v0 => gopkg.in/inf.v0 v0.9.1
 	gopkg.in/mcuadros/go-syslog.v2 => gopkg.in/mcuadros/go-syslog.v2 v2.2.1
 	gopkg.in/natefinch/lumberjack.v2 => gopkg.in/natefinch/lumberjack.v2 v2.0.0
+	gopkg.in/olivere/elastic.v2 => gopkg.in/olivere/elastic.v2 v2.0.12
```
Comment (liggitt) [inline go.mod]: this is only used by cadvisor elasticsearch storage (which we don't use). we should find a way to isolate storage and container integrations in cadvisor so we don't pull this into our dependency tree if we don't need to

File: go.mod [hunk 9]
```diff
@@ -186,14 +172,19 @@ replace (
...
 	github.com/NYTimes/gziphandler => github.com/NYTimes/gziphandler v0.0.0-20170623195520-56545f4a5d46
+	github.com/Nvveen/Gotty => github.com/Nvveen/Gotty v0.0.0-20120604004816-cd527374f1e5
 	github.com/OpenPeeDeeP/depguard => github.com/OpenPeeDeeP/depguard v1.0.1
 	github.com/PuerkitoBio/purell => github.com/PuerkitoBio/purell v1.1.1
 	github.com/PuerkitoBio/urlesc => github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578
-	github.com/Rican7/retry => github.com/Rican7/retry v0.1.0
+	github.com/Rican7/retry => github.com/Rican7/retry v0.1.1-0.20160712041035-272ad122d6e5
+	github.com/SeanDolphin/bqschema => github.com/SeanDolphin/bqschema v0.0.0-20150424181127-f92a08f515e1
+	github.com/Shopify/sarama => github.com/Shopify/sarama v1.8.0
+	github.com/Shopify/toxiproxy => github.com/Shopify/toxiproxy v2.1.4+incompatible
```
Comment (liggitt) [inline go.mod]: these pull in a lot of additional transitive dependencies and are only used by cadvisor kafka storage integration (which we don't use). we should find a way to isolate storage and container integrations in cadvisor so we don't pull this into our dependency tree if we don't need to

**Comment by liggitt** [comment]
@dims @dashpole since the last time we updated cadvisor, it added a go.mod file, which exposes all its transitive dependency version opinions, whether we make use of them or not. That means that all the container runtime (containerd, crio, docker, libcontainer, mesos, systemd) and storage integrations (bigquery, elasticsearch, influxdb, kafka, redis, statsd) that cadvisor has affect our dependency tree, even if we don't actually vendor the code in and link to it.

Unless there is an urgent reason to pull this in (e.g. fixing critical bugs), I'd like to see if we can isolate the cadvisor dependencies needed for the cadvisor cmd so downstream projects don't have to pull in the world like this. I'm experimenting with an approach locally in the cadvisor component and can sync up with you and @dashpole on it.

**Comment by dims** [comment]
@liggitt no rush other than the normal "try to get this stuff into master early in the cycle" so we can bake stuff in ci

**Comment by liggitt** [comment]
It looks like github.com/Microsoft/hcsshim@v0.8.7 also has a dependency on k8s.io/kubernetes@v1.13.0, which is problematic

https://github.com/microsoft/hcsshim/blob/master/go.mod#L36

File: go.mod [hunk 10]
```diff
@@ -59,41 +53,35 @@ require (
...
 	github.com/json-iterator/go v1.1.8
-	github.com/karrick/godirwalk v1.7.5 // indirect
 	github.com/libopenstorage/openstorage v1.0.0
 	github.com/lithammer/dedent v1.1.0
 	github.com/lpabon/godbc v0.1.1 // indirect
 	github.com/mattn/go-shellwords v1.0.5 // indirect
 	github.com/mesos/mesos-go v0.0.9 // indirect
 	github.com/miekg/dns v1.1.4
 	github.com/mindprince/gonvml v0.0.0-20190828220739-9ebdce4bb989 // indirect
-	github.com/mistifyio/go-zfs v2.1.1+incompatible // indirect
```
Comment (kazimsarikaya) [inline go.mod]: it fixes #87341

File: go.mod [hunk 11]
```diff
@@ -386,12 +402,12 @@ replace (
 	github.com/mattn/go-runewidth => github.com/mattn/go-runewidth v0.0.2
 	github.com/mattn/go-shellwords => github.com/mattn/go-shellwords v1.0.5
 	github.com/mattn/goveralls => github.com/mattn/goveralls v0.0.2
-	github.com/matttproud/golang_protobuf_extensions => github.com/matttproud/golang_protobuf_extensions v1.0.1
+	github.com/matttproud/golang_protobuf_extensions => github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
 	github.com/mesos/mesos-go => github.com/mesos/mesos-go v0.0.9
 	github.com/mholt/certmagic => github.com/mholt/certmagic v0.6.2-0.20190624175158-6a42ef9fe8c2
 	github.com/miekg/dns => github.com/miekg/dns v1.1.4
 	github.com/mindprince/gonvml => github.com/mindprince/gonvml v0.0.0-20190828220739-9ebdce4bb989
-	github.com/mistifyio/go-zfs => github.com/mistifyio/go-zfs v2.1.1+incompatible
```
Comment (kazimsarikaya) [inline go.mod]: it fixes #87341

File: go.mod [hunk 12]
```diff
@@ -386,12 +402,12 @@ replace (
...
+	github.com/mistifyio/go-zfs => github.com/mistifyio/go-zfs v2.1.2-0.20190413222219-f784269be439+incompatible
```
Comment (kazimsarikaya) [inline go.mod]: does #87341

File: go.mod (hunk 10 above)
Comment (kazimsarikaya) [inline go.mod]: it fixes #87341

**Review by kazimsarikaya — APPROVED** [review]
zfs-go related problems (#87341 ) are solved with this pr

**Comment by liggitt** [comment]
> It looks like github.com/Microsoft/hcsshim@v0.8.7 also has a declared dependency on k8s.io/kubernetes@v1.13.0, which is problematic
> 
> https://github.com/microsoft/hcsshim/blob/master/go.mod#L36

That is depending on cri-api, which is published to a staging module now. https://github.com/liggitt/hcsshim/commits/bump-k8s easily updates hcsshim to use the staging module, but that would mean we could not depend on that version of hcsshim (we do not allow k8s.io/kubernetes -> vendor -> staging cycles). The only way I see to unblock that is for k8s.io/cri-api to graduate from staging and be a standalone repo.

**Comment by dims** [comment]
ah fun error

```
ERROR(linux/386,linux/arm) vendor/github.com/opencontainers/selinux/go-selinux/selinux_linux.go:118:17: unix.SELINUX_MAGIC (untyped int constant 4185718668) overflows int32
exit status 1
!!! Type Check has failed. This may cause cross platform build failures.
!!! Please see https://git.k8s.io/kubernetes/test/typecheck for more information.
```

File: go.mod [hunk 13]
```diff
@@ -60,42 +53,35 @@ require (
...
 	github.com/mvdan/xurls v1.1.0
 	github.com/onsi/ginkgo v1.11.0
-	github.com/onsi/gomega v1.7.0
+	github.com/onsi/gomega v1.7.1
 	github.com/opencontainers/go-digest v1.0.0-rc1
-	github.com/opencontainers/image-spec v1.0.1 // indirect
-	github.com/opencontainers/runc v1.0.0-rc10
-	github.com/opencontainers/runtime-spec v1.0.0 // indirect
-	github.com/opencontainers/selinux v1.3.1-0.20190929122143-5215b1806f52
+	github.com/opencontainers/runc v1.0.0-rc8.0.20190906011214-a6606a7ae9d9
```
Comment (mattjmcnaughton) [inline go.mod]: Very possible I'm just misunderstanding how the release candidate tagging process works... but am I reading correctly that we're going from `v1.0.0-rc10` to `v1.0.0-rc8`? Would that not be a downgrade?

**Review by mattjmcnaughton — COMMENTED** [review]
W.r.t. `runc`, I think this looks good to me...

As far as I can tell, the main public facing impact is the addition of cgroups v2 support.

There is some deleted functionality, and some refactoring around where methods live... but that feels like implementation details.

I don't see any new `init` behavior or added dependencies or other things I'd consider suspicious.

File: go.mod (hunk 13 above)
Comment (dims) [reply go.mod]: good point, will revise this to keep the rc10

**Comment by brancz** [comment]
cc @dashpole

**Comment by liggitt** [comment]
looks like something weird happened to the kube-proxy go.sum in the `github.com/checkpoint-restore/go-criu` and `github.com/cilium/ebpf` commits (added 445 lines, then dropped them)

**Comment by dims** [comment]
@liggitt when we get the cadvisor stuff merged, i can do this again without touching the go-criu version, so we should be fine. Let's work on getting the `isolate` branch merged there.

**Comment by dims** [comment]
This too a while to get right. Especially happy with the changes in cadvisor to minimize the transitive dependencies and hence the churn in our go.* files. We can move to current cadvisor as-is to get other things in (like `dockerless` tag) please @liggitt and @dashpole

**Comment by liggitt** [comment]
just a couple comments:
1. we're not pinned to a go1.13 branch of golang/x/net, so floating to the preferred version would make sense to me
2. the influxdb commit can be dropped, since we just undo that change in the "drop unused pinned versions" commit

I made both of those changes in https://github.com/liggitt/kubernetes/commits/update-hcsshim-v0.8.7-and-containerd-v1.3.2 if you want to pick those up

**Comment by dims** [comment]
@liggitt Done with both items you listed. yep, thanks for sharing your branch it was helpful.

**Comment by mattjmcnaughton** [comment]
/retest

Exciting stuff :) Looking forward to rebasing https://github.com/kubernetes/kubernetes/pull/87746 on this once its merged.

**Comment by liggitt** [comment]
/hold

For release-blocking utils bump, to avoid merge conflicts

**Comment by dims** [comment]
/hold cancel

#89444 is merged and the cherry-pick #89449 is merged as well.

**Comment by BenTheElder** [comment]
I don't know why yet but this seemingly broke kind clusters on certain hosts (google's desktop linux distro and docker for mac I think so far).

https://github.com/kubernetes-sigs/kind/issues/1569#issuecomment-624933076

Not exactly expecting a revert based on that but ... FYI. And interested in any thoughts why this would be the relevant commit. Possibly kubelet unhealthy + api server does not come up ... not sure why yet.

**Comment by liggitt** [comment]
nothing leaps to mind, though I'd check the kubelet logs for hints first, given it is responsible for running the apiserver static pod and is the component most affected by this commit. Mind opening an issue to track/discuss?

**Comment by dims** [comment]
fyi, this problem was already fixed in  https://github.com/google/cadvisor/pull/2471

**Comment by dghubble** [comment]
On Fedora CoreOS (but not Flatcar), Kubelet can't start static pods beginning here (in v1.19.0-alpha.0, specifically 4274ea2c89dee24e4c188a71e8164b2a40d1e181) and present in v1.19.0-beta.1. Filed as https://github.com/kubernetes/kubernetes/issues/91795, but wanted to provide a 2nd case to show its broader than just kind and google's OS in case that's helpful.

**Comment by BenTheElder** [comment]
this is still present in https://github.com/kubernetes/kubernetes/issues/89859#issuecomment-624963169 as well

thanks for the details!

**Comment by dims** [comment]
thanks @dghubble that's helpful.

**Pull request merged**